lib/ASTGen/Sources/ASTGen/Decls.swift

```diff
@@ -1,162 +1,503 @@
+    return .decl(declAndDeclContext.decl)
   }
 
-  public func visit(_ node: FunctionDeclSyntax) -> ASTNode {
-    let staticLoc = bridgedSourceLoc(for: node)
-    let funcLoc = bridgedSourceLoc(for: node.funcKeyword)
-    let nameLoc = bridgedSourceLoc(for: node.identifier)
-    let rParamLoc = bridgedSourceLoc(for: node.signature.input.leftParen)
-    let lParamLoc = bridgedSourceLoc(for: node.signature.input.rightParen)
+  private func visitFunctionCodeBlock(_ block: CodeBlockSyntax, declContextAndDecl: BridgedDeclContextAndDecl) {
+    let oldDeclContext = declContext
+    declContext = declContextAndDecl.declContext
+    defer { declContext = oldDeclContext }
 
-    var nameText = node.identifier.text
-    let name = nameText.withBridgedString { bridgedName in
-      return ASTContext_getIdentifier(ctx, bridgedName)
+    AbstractFunctionDecl_setBody(declContextAndDecl.decl, self.visit(block).rawValue)
+  }
+}
+
+// MARK: - OperatorDecl
+
+extension BridgedOperatorFixity {
+  fileprivate init(tokenKind: TokenKind) {
+    switch tokenKind {
+    case .keyword(.infix): self = .infix
+    case .keyword(.prefix): self = .prefix
+    case .keyword(.postfix): self = .postfix
+    default: fatalError("Unknown operator fixity")
     }
+  }
+}
 
-    let returnType: ASTNode?
-    if let output = node.signature.output {
-      returnType = visit(output.returnType)
-    } else {
-      returnType = nil
+extension ASTGenVisitor {
+  func visit(_ node: OperatorDeclSyntax) -> ASTNode {
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
+    let (precedenceGroupName, precedenceGroupLoc) = self.bridgedIdentifierAndSourceLoc(
+      for: node.operatorPrecedenceAndTypes?.precedenceGroup
+    )
+
+    return .decl(
+      OperatorDecl_create(
+        self.ctx,
+        self.declContext,
+        BridgedOperatorFixity(tokenKind: node.fixity.tokenKind),
+        self.bridgedSourceLoc(for: node.operatorKeyword),
+        name,
+        nameLoc,
+        self.bridgedSourceLoc(for: node.operatorPrecedenceAndTypes?.colon),
+        precedenceGroupName,
+        precedenceGroupLoc
+      )
+    )
+  }
+}
+
+extension BridgedAssociativity {
+  fileprivate init(tokenKind: TokenKind) {
+    switch tokenKind {
+    case .keyword(.none): self = .none
+    case .keyword(.left): self = .left
+    case .keyword(.right): self = .right
+    default: fatalError("Unknown precedence group associativity value")
     }
+  }
+}
 
-    let params = node.signature.input.parameterList.map { visit($0).rawValue }
-    let out = params.withBridgedArrayRef { ref in
-      FuncDecl_create(
-        ctx, staticLoc, false, funcLoc, name, nameLoc, false, nil, false, nil, rParamLoc, ref,
-        lParamLoc,
-        returnType?.rawValue, declContext)
+extension ASTGenVisitor {
+  func visit(_ node: PrecedenceGroupDeclSyntax) -> ASTNode {
+    let (name, nameLoc) = self.bridgedIdentifierAndSourceLoc(for: node.identifier)
+
+    struct PrecedenceGroupBody {
```

Comment:
For precedence groups and operators, I wonder if we should instead build on top of the `SwiftOperators` library, which has a mapping from the syntax tree into a more semantic representation (`Operator` and `PrecedenceGroup`).

Reply:
Apparently the inits from `SwiftOperators` assume the nodes are valid, so we would have to do most of the same work in `ASTGen` to prevalidate the nodes and diagnose issues before calling these inits, like switching on the fixity of an operator or iterating over the attributes of a precedence group. Do you believe there is something to gain from building on top of `SwiftOperators` regardless?